## Exploring eBay Car Sales Data
### Guided Project - Dataquest

We are going to work with a dataset of used cars from *eBay Kleinanzeigen*. The aim of this project is to **clean the data** and **analyze** the included used car listings.

The dataset was originally scraped and uploaded to Kaggle by user orgesleka and the original dataset isn't available on Kaggle anymore. We can download it from [here](https://data.world/data-society/used-cars-data) with some modifications now:

- 50,000 data points from the full dataset.
- Dataset was dirtied to resemble what you would expect from a scraped dataset.

The data dictionary provided with data is as follows:

|Column | Meaning|
|:-------|:--------|
|dateCrawled | When this ad was first crawled. All field-values are taken from this date|
name | Name of the car.
seller | Whether the seller is private or a dealer.
offerType | The type of listing
price | The price on the ad to sell the car.
abtest | Whether the listing is included in an A/B test.
vehicleType | The vehicle Type.
yearOfRegistration | The year in which the car was first registered.
gearbox | The transmission type.
powerPS | The power of the car in PS.
model | The car model name.
kilometer | How many kilometers the car has driven.
monthOfRegistration | The month in which the car was first registered.
fuelType | What type of fuel the car uses.
brand | The brand of the car.
notRepairedDamage | If the car has a damage which is not yet repaired.
dateCreated | The date on which the eBay listing was created.
nrOfPictures | The number of pictures in the ad.
postalCode | The postal code for the location of the vehicle.
lastSeenOnline | When the crawler saw this ad last online.



In [1]:
#import NumPy and Pandas modules
import numpy as np
import pandas as pd

#import csv file with pandas
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [2]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

In [3]:
#Information about the dataset
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

We can see there are 20 columns. The columns name use camelcase instead of Python's preferred snakecase.

Information about dates in `dateCrawled`,`dateCreated` and `lastSeen` (%Y-%m-%d %H:%M:%s) seem to be OK.

We are going to clean `name` as spaces are replaced by _ and we are going to capitalize the first letter of each word with .title().

`Seller`, `offerType`, `abtest`, `gearbox` and `notRepairedDamage` columns have an string as value. Some of this information is in German (seller,gearbox,notRepairedDamage), we can think about translating this info into English.

`Price` and `kilometer` columns are int. It can be more interesting to have a float as type.
`yearOfRegistration`, `nrOfPictures` and `postalCode` colums seems to be OK (int).
We must check `monthOfRegistration` and `powerPS` column, there is a 0 in the first row!

There are some NaN as `vehicleType` and `model`.

In [4]:
#Change column names from camelCase to snake_case
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
new_col_names = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.columns = new_col_names
print(autos.columns)
autos.head()

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

The 20 columns name are using snake_case (Python's convention) instead of camelCase now.

Let's start exploring data now

In [6]:
autos.describe()

price  registration_year       power_ps      kilometer  \
count  3.715280e+05      371528.000000  371528.000000  371528.000000   
mean   1.729514e+04        2004.577997     115.549477  125618.688228   
std    3.587954e+06          92.866598     192.139578   40112.337051   
min    0.000000e+00        1000.000000       0.000000    5000.000000   
25%    1.150000e+03        1999.000000      70.000000  125000.000000   
50%    2.950000e+03        2003.000000     105.000000  150000.000000   
75%    7.200000e+03        2008.000000     150.000000  150000.000000   
max    2.147484e+09        9999.000000   20000.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count       371528.000000        371528.0  371528.00000  
mean             5.734445             0.0   50820.66764  
std              3.712412             0.0   25799.08247  
min              0.000000             0.0    1067.00000  
25%              3.000000             0.0   30459.00000  
50%              6.000000             0.0   49610.00000  
75%              9.000000             0.0   71546.00000  
max             12.000000             0.0   99998.00000

The describe() method gives information about the numeric columns of the dataset.

We can clearly observe some data to be dropped in `registration_month` (there are 0 as min when it is supposed to be from 1 to 12), `registration_year` (min = 1000 and max = 9999, time travels have not been invented yet) or `power_ps` equals to 0.

Prices over 2 billions? Suspicious...

In [7]:
print(autos["seller"].value_counts(),'\n')
print(autos["offer_type"].value_counts(),'\n')
print(autos["abtest"].value_counts(),'\n')
print(autos["gearbox"].value_counts(),'\n')
print(autos["unrepaired_damage"].value_counts(),'\n')
print(autos["vehicle_type"].value_counts(),'\n')
print(autos["fuel_type"].value_counts(),'\n')

privat        371525
gewerblich         3
Name: seller, dtype: int64 

Angebot    371516
Gesuch         12
Name: offer_type, dtype: int64 

test       192585
control    178943
Name: abtest, dtype: int64 

manuell      274214
automatik     77105
Name: gearbox, dtype: int64 

nein    263182
ja       36286
Name: unrepaired_damage, dtype: int64 

limousine     95894
kleinwagen    80023
kombi         67564
bus           30201
cabrio        22898
coupe         19015
suv           14707
andere         3357
Name: vehicle_type, dtype: int64 

benzin     223857
diesel     107746
lpg          5378
cng           571
hybrid        278
andere        208
elektro       104
Name: fuel_type, dtype: int64 



In [8]:
#Translation into English
autos.loc[autos["seller"] == "privat","seller"] = "Private"
autos.loc[autos["seller"] == "gewerblich","seller"] = "Commercial"

autos.loc[autos["offer_type"] == "Angebot","offer_type"] = "Offer"
autos.loc[autos["offer_type"] == "Gesuch","offer_type"] = "Request"

autos.loc[autos["abtest"] == "test","abtest"] = "Test"
autos.loc[autos["abtest"] == "control","abtest"] = "Control"

autos.loc[autos["gearbox"] == "manuell","gearbox"] = "Manually"
autos.loc[autos["gearbox"] == "automatik","gearbox"] = "Automatic"

autos.loc[autos["unrepaired_damage"] == "nein","unrepaired_damage"] = "No"
autos.loc[autos["unrepaired_damage"] == "ja","unrepaired_damage"] = "Yes"

autos.loc[autos["vehicle_type"] == "limousine","vehicle_type"] = "Limousine"
autos.loc[autos["vehicle_type"] == "kleinwagen","vehicle_type"] = "Compact"
autos.loc[autos["vehicle_type"] == "kombi","vehicle_type"] = "Wagon"
autos.loc[autos["vehicle_type"] == "bus","vehicle_type"] = "Bus"
autos.loc[autos["vehicle_type"] == "cabrio","vehicle_type"] = "Cabriolet"
autos.loc[autos["vehicle_type"] == "coupe","vehicle_type"] = "Coupe"
autos.loc[autos["vehicle_type"] == "suv","vehicle_type"] = "SUV"
autos.loc[autos["vehicle_type"] == "andere","vehicle_type"] = "Other"

autos.loc[autos["fuel_type"] == "benzin","fuel_type"] = "Gasoline"
autos.loc[autos["fuel_type"] == "diesel","fuel_type"] = "Diesel"
autos.loc[autos["fuel_type"] == "lpg","fuel_type"] = "LPG"
autos.loc[autos["fuel_type"] == "cng","fuel_type"] = "CNG"
autos.loc[autos["fuel_type"] == "hybrid","fuel_type"] = "Hybrid"
autos.loc[autos["fuel_type"] == "andere","fuel_type"] = "Other"
autos.loc[autos["fuel_type"] == "elektro","fuel_type"] = "Electric"


print(autos["seller"].value_counts(),'\n')
print(autos["offer_type"].value_counts(),'\n')
print(autos["abtest"].value_counts(),'\n')
print(autos["gearbox"].value_counts(),'\n')
print(autos["unrepaired_damage"].value_counts(),'\n')
print(autos["vehicle_type"].value_counts(),'\n')
print(autos["fuel_type"].value_counts(),'\n')

Private       371525
Commercial         3
Name: seller, dtype: int64 

Offer      371516
Request        12
Name: offer_type, dtype: int64 

Test       192585
Control    178943
Name: abtest, dtype: int64 

Manually     274214
Automatic     77105
Name: gearbox, dtype: int64 

No     263182
Yes     36286
Name: unrepaired_damage, dtype: int64 

Limousine    95894
Compact      80023
Wagon        67564
Bus          30201
Cabriolet    22898
Coupe        19015
SUV          14707
Other         3357
Name: vehicle_type, dtype: int64 

Gasoline    223857
Diesel      107746
LPG           5378
CNG            571
Hybrid         278
Other          208
Electric       104
Name: fuel_type, dtype: int64 



In [9]:
#Cleaning car names
#We are going to clean name as spaces are replaced by _ and we are going to capitalize the first letter of each word with .title().

autos["name"] = autos["name"].str.replace('_',' ').str.title()


#Cleaning model and brand car
autos["model"] = autos["model"].str.replace('_',' ').str.title()
autos["brand"] = autos["brand"].str.replace('_',' ').str.title()

autos.head()

date_crawled                            name   seller offer_type  \
0  2016-03-24 11:52:17                      Golf 3 1.6  Private      Offer   
1  2016-03-24 10:58:45            A5 Sportback 2.7 Tdi  Private      Offer   
2  2016-03-14 12:52:21  Jeep Grand Cherokee "Overland"  Private      Offer   
3  2016-03-17 16:54:04              Golf 4 1 4  3Türer  Private      Offer   
4  2016-03-31 17:25:20  Skoda Fabia 1.4 Tdi Pd Classic  Private      Offer   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   Test          NaN               1993   Manually         0   Golf   
1  18300   Test        Coupe               2011   Manually       190    NaN   
2   9800   Test          SUV               2004  Automatic       163  Grand   
3   1500   Test      Compact               2001   Manually        75   Golf   
4   3600   Test      Compact               2008   Manually        69  Fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0  Gasoline  Volkswagen               NaN   
1     125000                   5    Diesel        Audi               Yes   
2     125000                   8    Diesel        Jeep               NaN   
3     150000                   6  Gasoline  Volkswagen                No   
4      90000                   7    Diesel       Skoda                No   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

Let's study price and kilometer column

In [10]:
print('Total of unique rows of price column: ', autos["price"].unique().shape[0],'\n')
print('Description: \n', autos["price"].describe(),'\n')
print('Highest prices: \n',autos["price"].value_counts().sort_index(ascending=False).head(10),'\n') #high prices
print('Lowest prices: \n',autos["price"].value_counts().sort_index(ascending=True).head(10),'\n') #low prices

print('Total of unique rows of kilometer column: ', autos["kilometer"].unique().shape[0],'\n')
print('Description: \n', autos["kilometer"].describe(),'\n')
print('Highest kilometers: \n',autos["kilometer"].value_counts().sort_index(ascending=False).head(10),'\n') #high kilometer
print('Lowest kilometers: \n',autos["kilometer"].value_counts().sort_index(ascending=True).head(10),'\n') #low kilometer

Total of unique rows of price column:  5597 

Description: 
 count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64 

Highest prices: 
 2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
Name: price, dtype: int64 

Lowest prices: 
 0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
Name: price, dtype: int64 

Total of unique rows of kilometer column:  13 

Description: 
 count    371528.000000
mean     125618.688228
std       40112.337051
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: kilometer, dtype: float64 

Highest kilometers: 
 150000    240797
125000     38067
100000     1592

`kilometer` seems to have correct data but it is not the case for `price`.
We are going to remove rows with prices over 100 000 euros and less than 200 euros.

In [11]:
autos = autos[autos["price"].between(200,100000)]

print('Total rows of price column: ', autos["price"].unique().shape[0],'\n')
print('Description: \n', autos["price"].describe(),'\n')
print('Highest prices: \n',autos["price"].value_counts().sort_index(ascending=False).head(10),'\n') #high prices
print('Lowest prices: \n',autos["price"].value_counts().sort_index(ascending=True).head(10),'\n') #low prices

Total rows of price column:  5271 

Description: 
 count    353655.000000
mean       5882.955072
std        7581.685489
min         200.000000
25%        1300.000000
50%        3200.000000
75%        7500.000000
max      100000.000000
Name: price, dtype: float64 

Highest prices: 
 100000     6
99999     13
99990      2
99911      1
99900      6
99500      2
99000      8
98900      1
98500      2
98430      1
Name: price, dtype: int64 

Lowest prices: 
 200    1986
202       1
205       2
209       2
210      19
211       2
215      11
217       1
219       7
220     157
Name: price, dtype: int64 



It is time to analyse data about dates:

- `date_crawled`: added by the crawler - string (%Y-%m-%d %H:%M:%s)
- `last_seen`: added by the crawler - string (%Y-%m-%d %H:%M:%s)
- `ad_created`: from the website - string (%Y-%m-%d %H:%M:%s)
- `registration_month`: from the website - int
- `registration_year`: from the website - int

In [12]:
#Distribution:
print(autos[['date_crawled','ad_created','last_seen']][0:5],'\n')
print('date_crawled: \n', autos['date_crawled'].str[:10].value_counts().sort_index(),'\n')
print('ad_created: \n', autos['ad_created'].str[:10].value_counts().sort_index(),'\n')
print('last_seen: \n', autos['last_seen'].str[:10].value_counts().sort_index(),'\n')

          date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21 

date_crawled: 
 2016-03-05     9066
2016-03-06     5116
2016-03-07    12584
2016-03-08    11828
2016-03-09    12060
2016-03-10    11543
2016-03-11    11572
2016-03-12    12867
2016-03-13     5601
2016-03-14    12822
2016-03-15    11797
2016-03-16    10679
2016-03-17    11181
2016-03-18     4637
2016-03-19    12448
2016-03-20    12843
2016-03-21    12594
2016-03-22    11459
2016-03-23    11317
2016-03-24    10535
2016-03-25    11593
2016-03-26    11340
2016-03-27    10701
2016-03-28    12416
2016-03-29    12043
2016-03-30    11864
2016-03-31    11268
2016-04-01    12111
2016-04-02    12465
2016-04-03  

In [13]:
#Percentage
print('date_crawled: \n',autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()*100,'\n')
print('ad_created: \n',autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()*100,'\n')
print('last_seen: \n',autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()*100,'\n')

date_crawled: 
 2016-03-05    2.563515
2016-03-06    1.446608
2016-03-07    3.558270
2016-03-08    3.344502
2016-03-09    3.410103
2016-03-10    3.263915
2016-03-11    3.272115
2016-03-12    3.638292
2016-03-13    1.583747
2016-03-14    3.625567
2016-03-15    3.335737
2016-03-16    3.019610
2016-03-17    3.161556
2016-03-18    1.311165
2016-03-19    3.519815
2016-03-20    3.631505
2016-03-21    3.561098
2016-03-22    3.240163
2016-03-23    3.200011
2016-03-24    2.978892
2016-03-25    3.278053
2016-03-26    3.206515
2016-03-27    3.025830
2016-03-28    3.510766
2016-03-29    3.405296
2016-03-30    3.354682
2016-03-31    3.186156
2016-04-01    3.424524
2016-04-02    3.524621
2016-04-03    3.898715
2016-04-04    3.770341
2016-04-05    1.276951
2016-04-06    0.311886
2016-04-07    0.159477
Name: date_crawled, dtype: float64 

ad_created: 
 2014-03-10    0.000283
2015-03-20    0.000283
2015-06-11    0.000283
2015-06-18    0.000283
2015-08-07    0.000283
                ...   
2016-04-03   

In [14]:
#Description about registration_month and registration_year
print(autos["registration_month"].describe(),'\n')
print(autos["registration_year"].describe(),'\n')

count    353655.000000
mean          5.840463
std           3.666600
min           0.000000
25%           3.000000
50%           6.000000
75%           9.000000
max          12.000000
Name: registration_month, dtype: float64 

count    353655.000000
mean       2004.321924
std          70.565849
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64 



We can see there are some errors:
- Month are supposed to be between 1 and 12 but we observe we can find a month 0.
- Year cannot be 1000 or 9999. [1900-2016] is a more accurate period.
    
We are going to analyse the value_counts for both month and year.

In [15]:
print('Months: \n',autos['registration_month'].value_counts().sort_index(),'\n')
print('Years: \n',autos['registration_year'].value_counts().sort_index(),'\n')

Months: 
 0     30404
1     23544
2     21625
3     35094
4     29905
5     29600
6     32103
7     28072
8     22992
9     24357
10    26574
11    24818
12    24567
Name: registration_month, dtype: int64 

Years: 
 1000    20
1001     1
1039     1
1111     1
1234     3
        ..
8500     1
8888     1
9000     4
9450     1
9999    12
Name: registration_year, Length: 140, dtype: int64 



In [16]:
print('Total count year: ', autos["registration_year"].shape[0])
print('Total count year < 1900 or year > 2016: ',autos.loc[(autos['registration_year']< 1900) | (autos['registration_year']> 2016),"registration_year"].shape[0])
print('Total count 1900 < year < 2016: ',autos.loc[autos['registration_year'].between(1900,2016),"registration_year"].shape[0])

Total count year:  353655
Total count year < 1900 or year > 2016:  13796
Total count 1900 < year < 2016:  339859


There are 13796 rows with a wrong year. That represents the 3.9% of total data. We decide to delete them.

In [17]:
autos = autos[autos["registration_year"].between(1900,2016)]
print('Registration years (%): \n',autos['registration_year'].value_counts(normalize=True)*100,'\n')

Registration years (%): 
 2000    6.511818
1999    6.345867
2005    6.325270
2006    5.871847
2001    5.757976
          ...   
1942    0.000588
1927    0.000588
1940    0.000588
1925    0.000294
1911    0.000294
Name: registration_year, Length: 94, dtype: float64 



Most part of cars were registred in 2000's. We can do the same study with the registration month.

In [18]:
print('Total count month: ', autos["registration_month"].shape[0])
print('Total count month = 0: ',autos.loc[(autos['registration_month'] == 0),"registration_month"].shape[0])
print('Total count 1 < month < 12: ',autos.loc[autos['registration_month'].between(1,12),"registration_month"].shape[0])

Total count month:  339859
Total count month = 0:  27032
Total count 1 < month < 12:  312827


There are 27032 rows with a wrong month. That represents the 7.95% of total data. We decide to let the data included as we are not interested in this data for now.

Let's check the brand column now:

In [19]:
#Analysis about price / brand on average
brands = autos["brand"].unique()

brands_price = {}

for b in brands:
    avg_price = sum(autos.loc[autos["brand"] == b,"price"]) / len(autos.loc[autos["brand"] == b,"price"])
    brands_price[b] = str(round(avg_price,2))+'€'

print('Price per brand on average: ')
brand_price_serie = pd.Series(brands_price)
print(brand_price_serie)
    

Price per brand on average: 
Volkswagen         5489.27€
Audi               9109.36€
Jeep              11287.46€
Skoda              6546.46€
Bmw                 8462.9€
Peugeot            3324.65€
Ford               3827.84€
Mazda              4148.32€
Nissan             4838.04€
Renault            2513.07€
Mercedes Benz      8423.97€
Seat               4609.22€
Honda              4037.62€
Fiat               2988.64€
Opel               3067.89€
Mini              10114.58€
Smart              3649.97€
Hyundai            5624.64€
Alfa Romeo         4378.23€
Subaru             4568.63€
Volvo              5297.55€
Mitsubishi         3454.27€
Kia                5906.13€
Suzuki             4104.27€
Lancia             3430.36€
Porsche           34331.69€
Citroen            3816.48€
Sonstige Autos    12247.55€
Toyota             5323.97€
Chevrolet          7146.44€
Dacia               5943.3€
Daihatsu           1881.05€
Chrysler           4192.49€
Jaguar            12235.03€
Daewoo             

We can see Porsche (34331.69€) and Land Rover (16697.39€) are the most expensive cars and the cheapest ones are Daewo (1067.71€).

- Porsche :  34331.69€
- Land Rover :  16697.39€
- Sonstige Autos :  12247.55€
- Jaguar :  12235.03€
- Jeep :  11287.46€
- ...
- Opel :  3067.89€
- Fiat :  2988.64€
- Renault :  2513.07€
- Trabant :  2063.82€
- Daihatsu :  1881.05€
- Rover :  1674.77€
- Daewoo :  1067.71€

The mean is 5882.95 euros/car



















In [20]:
#Analysis about km / brand on average
brands_km = {}

for b in brands:
    avg_km = sum(autos.loc[autos["brand"] == b,"kilometer"]) / len(autos.loc[autos["brand"] == b,"kilometer"])
    brands_km[b] = str(round(avg_km,2))+' Km'

print('Km per brand on average: ')
brand_km_serie = pd.Series(brands_km)
print(brand_km_serie)

Km per brand on average: 
Volkswagen        128444.21 Km
Audi              129591.76 Km
Jeep               120675.5 Km
Skoda             113600.52 Km
Bmw               133052.44 Km
Peugeot           124332.39 Km
Ford              123586.69 Km
Mazda             125625.24 Km
Nissan            119079.61 Km
Renault           127745.47 Km
Mercedes Benz     130868.52 Km
Seat               120482.8 Km
Honda             125853.28 Km
Fiat              116331.75 Km
Opel              128618.24 Km
Mini               93388.34 Km
Smart              99694.58 Km
Hyundai            104064.6 Km
Alfa Romeo        129309.38 Km
Subaru            126171.88 Km
Volvo             138263.16 Km
Mitsubishi        126902.02 Km
Kia               109175.08 Km
Suzuki            106326.34 Km
Lancia            123321.76 Km
Porsche           103626.26 Km
Citroen           120144.85 Km
Sonstige Autos     89904.98 Km
Toyota            117290.64 Km
Chevrolet         100446.94 Km
Dacia              85986.08 Km
Daihatsu     

We can see Saab (141590.46 Km) and Volvo (138263.16 Km) are the cars with more kilometers on average. Dacia (85986.08 Km), Trabant (55424.11 Km) and Lada (74195.12 Km) have less kilometers on average.

The mean is 125618.69 km/car


In [21]:
#Create dataframe for brands
branddf = pd.DataFrame(brand_price_serie, columns=['mean_price'])

branddf["mean_km"] = brand_km_serie

branddf

mean_price       mean_km
Volkswagen       5489.27€  128444.21 Km
Audi             9109.36€  129591.76 Km
Jeep            11287.46€   120675.5 Km
Skoda            6546.46€  113600.52 Km
Bmw               8462.9€  133052.44 Km
Peugeot          3324.65€  124332.39 Km
Ford             3827.84€  123586.69 Km
Mazda            4148.32€  125625.24 Km
Nissan           4838.04€  119079.61 Km
Renault          2513.07€  127745.47 Km
Mercedes Benz    8423.97€  130868.52 Km
Seat             4609.22€   120482.8 Km
Honda            4037.62€  125853.28 Km
Fiat             2988.64€  116331.75 Km
Opel             3067.89€  128618.24 Km
Mini            10114.58€   93388.34 Km
Smart            3649.97€   99694.58 Km
Hyundai          5624.64€   104064.6 Km
Alfa Romeo       4378.23€  129309.38 Km
Subaru           4568.63€  126171.88 Km
Volvo            5297.55€  138263.16 Km
Mitsubishi       3454.27€  126902.02 Km
Kia              5906.13€  109175.08 Km
Suzuki           4104.27€  106326.34 Km
Lancia           3430.36€  123321.76 Km
Porsche         34331.69€  103626.26 Km
Citroen          3816.48€  120144.85 Km
Sonstige Autos  12247.55€   89904.98 Km
Toyota           5323.97€  117290.64 Km
Chevrolet        7146.44€  100446.94 Km
Dacia             5943.3€   85986.08 Km
Daihatsu         1881.05€  118252.84 Km
Chrysler         4192.49€  134407.29 Km
Jaguar          12235.03€  124428.33 Km
Daewoo           1067.71€  123347.11 Km
Rover            1674.77€  135535.71 Km
Saab             3994.21€  141590.46 Km
Land Rover      16697.39€  119532.09 Km
Trabant          2063.82€   55424.11 Km
Lada             3221.11€   74195.12 Km